lightning/src/ln/wire.rs

```diff
@@ -238,6 +263,18 @@ pub trait Encode {
 	}
 }
 
+/// A message that has an associated type id.
+pub trait TypedMessage {
+	/// The type id for the implementing message.
+	fn msg_type(&self) -> u16;
+}
+
+impl<T> TypedMessage for T where T: Encode {
```

Comment:
Note that you can avoid making `Encode` public with a dummy wrapper module - the following compiles (but obviously use a better module name and fix the indentation).
```
@@ -251,2 +251,4 @@ pub fn write<M: TypedMessage + Writeable, W: Writer>(message: &M, buffer: &mut W
 
+pub(crate) mod lulz {
+use super::*;
 /// Defines a type-identified encoding for sending messages over the wire.
@@ -264,2 +266,4 @@ pub trait Encode {
 }
+}
+pub(crate) use self::lulz::Encode;
 
```

Reply:
That's a nice trick thanks!